launcher/ui/pages/instance/ServersPage.cpp

```diff
@@ -112,11 +125,42 @@ struct Server {
     bool m_checked = false;
     bool m_up = false;
     QString m_motd;  // https://mctools.org/motd-creator
-    int m_ping = 0;
-    int m_currentPlayers = 0;
+    std::optional<int> m_currentPlayers; // nullopt if not calculated/calculating
     int m_maxPlayers = 0;
 };
 
+class ServerPingTask : public Task {
+    Q_OBJECT
+  public:
+    explicit ServerPingTask(QObject* parent, Server &server) : Task(parent), m_server(server) {}
+    ~ServerPingTask() override = default;
+
+
+  protected:
+    virtual void executeTask() override {
+        qDebug() << "Querying status of " << m_server.m_address;
+        auto [domain, port] = m_server.splitAddress();
+        McResolver *resolver = new McResolver(nullptr, domain, port);
+        QObject::connect(resolver, &McResolver::succeed, [=](QString ip, int port) {
+            resolver->deleteLater();
+            qDebug() << "Resolved Addresse for" << domain << ": " << ip << ":" << port;
+            McClient client(nullptr, domain, ip, port);
+
+            try {
+                int online = client.getOnlinePlayers();
+                qDebug() << "Online players: " << online;
+                m_server.m_currentPlayers = online;
+            } catch(const Exception& e) {
+                qDebug() << "Failed to get online players: " << e.cause();
+            }
+        });
+        resolver->ping();
+    }
+    
+  private:
+    Server &m_server;
```

Comment:
This may cause a crash

Reply:
Could you explain your reasoning ? Using `m_server` in the callback is indeed causing a crash right now (only sometimes, so I assume it's a race condition), but I cannot figure out why that would happen

Reply:
you pass a reference that reference is valid as long as the original reference is valid.
So once the server object that you passed to task is destructed this variable will not point to a correct memory address.
So what can cause the crash is as follows:
- the server is created on the ServerPage(and maybe associated with it)
- you pass the reference to this task(right now a pointer)
- task is executed async
- the serverPage is destroyed and together with it the server object
- but because you used a task pointer there are a few cases where they remain until the task is finished( e.g. you passed to the lambda a reference to that pointer to deleteLater)
- the task still has an invalid reference

PS. This is a speculation and may not be valid for you case but this is a recent crash caused by the capture of the tasks as a shared pointer.

Reply:
This is not what's happening in this specific case since I don't leave the server tab, but that's still something I'll have to take into consideration, thanks !

Reply:
Ok, found out what's happening: the `ServerPingTask` object is being deleted before the task finishes, probably because `Task::Ptr()` immediately calls `deleteLater()` on the object without waiting for it to finish.

I'll try to figure out a solution to this, probably just by linking the finished() signal of the task to deleteLater()

Reply:
In the end, the reason was that `job` was deleted a the end of the function call (what you said in https://github.com/PrismLauncher/PrismLauncher/pull/3112#discussion_r1846914296), causing the tasks to be destroyed, even though the they were not completed

Reply:
I still want this to be something else not a reference: is a bad idea to modify a reference async and there is no way to check if is still valid

Reply:
Done ! With 7c61fec8e0e753de1b7215e45b0a4272551ffac4 and fef8ee2d1b4f866437b30ec4ea8fd5cae2426e9c
Just tried, and closing the popup indeed stops `server` from being modified once the task has finished